### ISSUE TITLE
Support/approach for sliding window/multiple snapshots in time

### ISSUE THREAD

**Issue by rquintino** [FeatureLabs/featuretools#100]
Hi there! 
(first of all huge thx for dfs, vision & tools, superb work)

My question, the predict_next_purchase sample uses a single cut_off time right? But doesnt that remove a lot of data that could help with the purchase prediction? and  we're only using a single day for reference right?

only this data/users -> **_"Using users who had acivity during training_window days before the cutoff_time, we look to see if they purchase the product in the prediction_window."_**

I would like to use all data in a single final ml table for the models. Is there support to have the cut off being a sliding window (ex: for each customer) of features from last x days, predicting purchase (yes/no) up x days in the future.  So each customer would appear multiple times, depending on choosen sliding  window.

Think it's a tipical pattern in predicting future events (predictive maintenance, churn, healthcare). Usually applies to any kind of event prediction. (ex: for every user, machine, predict probability of event E for the next x days for a specific point in time, obv the training the dataset has proper timestamps so that we can "recalculate" feature values for user/machine up to at any point in time)

The dataset becomes non IID obv, some cautions apply.

Makes sense? What's the approach to use DFS with these scenarios?
thx!

**Comment by kmax12** [comment]
Thanks for the question, @rquintino.

Featuretools is designed to handle this exact use case. Read about the `cutoff_time` time argument in our [documentation](https://docs.featuretools.com/automated_feature_engineering/handling_time.html).

It explains how you can pass in multiple time different times to calculate features in a single call to `dfs` or `calculate_feature_matrix`. 

The cutoff time is used to specify the last time you should use data (because your label comes from after this time). By default, featuretools will use all available data from before the cutoff_time. However, if provide the `training_window` parameter you can specify a using less data. 

Basically, combining those two parameter you can achieve what you are describing. 

Let us know if the above is helpful or if you have additional questions.

Thanks!

**Comment by rquintino** [comment]
Hi Max, thanks! (honored really! :) )

yes, read it and was looking into the cutoff_time details. Issue was mainly  the purchase sample does not use this "sliding pattern" right? If so would be interesting a sliding window version for the sample, otherwise don't see much relevance on training a superv model in single cutoff?

(purchase sample single cutoff, notebook /utills.py)

label_times = utils.make_labels(es=es,
                                product_name = "Banana",
                                cutoff_time = pd.Timestamp('March 15, 2015'),
                                ....)

def make_labels(es, training_window, **cutoff_time**,
                product_name, prediction_window):
   ...
   **labels["cutoff_time"] = cutoff_time**
   ....

**Comment by rquintino** [comment]
related: in case you want to merge different training_windows in the same feature set, we would have to call dfs again? See its a dictionary but think to support different entity sets not multiple time windows?(ex: 90 days features+180 days features+36 hours features)

**Comment by kmax12** [comment]
Correct, that demo doesn't use that sliding pattern.

If you'd like, you can try to modify the `make_labels` function yourself to create multiple label_times per customer. all you should have to do is modify that one function and the rest will work the same!

Feel free to fork and give it a shot! I'd be happy to take a look if you do :)

**Comment by kmax12** [comment]
correct, all rows in a given call need the same `training_window`. I cannot think of use case for a feature matrix to be calculated with different `training_window`s

**Comment by rquintino** [comment]
will look into the make_labels. thx! probably an easy path can be just map/aggregate make_labels output for a few different sliding cutoff times. Not performance optimized though :), but should work, we'll take a look!
thx!

**Comment by rquintino** [comment]
question: regarding the purchase sample and adding multiple cutoff times.

How should we provide the (possible) multiple cutofftimes for user in dfs function ? Does it support receiving multiple cutoff_times for same instance_id? (thinking on having a cutoff each month day 20 for that dataset, so users can have more than one cutoff)

Or should we model this, adding new entity and use that one as target (in that case, new entity would be "snaphot")

huge thx Max!

**Comment by kmax12** [comment]
>How should we provide the (possible) multiple cutofftimes for user in dfs function ? Does it support receiving multiple cutoff_times for same instance_id? (thinking on having a cutoff each month day 20 for that dataset, so users can have more than one cutoff)

yep! that's one of the ways featuretools is so powerful :)

**Comment by TuSKan** [comment]
Hi @kmax12 , I have the same question.
You sad: _correct, all rows in a given call need the same training_window. I cannot think of use case for a feature matrix to be calculated with different training_windows_
Thinking on this way: I don't know the optimum window size (_training_window_) could be for example 6, 3 or 1 months before cutoff_time. I would like to test all this possibilities.
How can I create _MEAN_last_6months(value)_, _MEAN_last_3months(value)_ and _MEAN_last_1months(value)_ using **Featuretools** for the same cutoff_time ??

**Comment by kmax12** [comment]
@TuSKan that makes sense. One thing you can do is run `ft.calculate_feature_matrix` multiple times with different training windows. Then, you can either train models on the different feature matrices independently or merge them into one feature matrix.

**Comment by jrkinley** [comment]
I also have a requirement for multiple `training_window` durations. For my use case, predicting the outcome is based on a variety of signals. Some signals are sparse and need to be analysed over many months or years (e.g. purchase history). Whereas other signals are more intent-based and need to be analysed over days (e.g. browsing history or communications). The later signals are better predictors when analysed over shorter windows as they represent a recent change in behaviour or signal of intent towards the outcome I'm trying to predict.

In this scenario, do you recommend separating the features into two entity sets, running `calculate_feature_matrix` with different training windows on each, and subsequently combining the results into a single matrix? Assuming I have a common `cutoff_time` dataframe the resulting matrices should have the same index and thus be easy to join.

**Comment by kmax12** [comment]
@jrkinley That is a approximately the approach I would take. However, instead of having two entity sets, I would recommend using one entity set, but making two different calls to `ft.dfs` with `features_only==True`. You can use different primitives and different settings to ignore_variables in those `ft.dfs` calls to get two different lists of features definitions.

Then, you can call `ft.calculate_feature_matrix(..)` on each of those lists feature definition with different parameters for `training_window` and join. 

This is the approach we take in [this stackoverflow answer](https://stackoverflow.com/a/52593818/8964531), with the exception that we only create one feature list

**Comment by jrkinley** [comment]
@kmax12 super helpful, thanks.

**Comment by kmax12** [comment]
This appears to be resolved. If there are any further questions or issues, please post.
